src/py21cmfast/_memory.py

```diff
@@ -0,0 +1,706 @@
+
+    """Memory usage within IonisationBox.c"""
+    hii_kspace_num_pixels = (float(user_params.HII_DIM) / 2.0 + 1.0) * (
+        user_params.HII_DIM
+    ) ** 2
+
+    # deltax_unfiltered, delta_unfiltered_original, deltax_filtered
+    num_c_boxes = 3.0
+
+    if flag_options.USE_MINI_HALOS:
+        # prev_deltax_unfiltered, prev_deltax_filtered
+        num_c_boxes += 2.0
+
+    if flag_options.USE_TS_FLUCT:
+        # xe_unfiltered, xe_filtered
+        num_c_boxes += 2.0
+
+    if flag_options.INHOMO_RECO:
+        # N_rec_unfiltered, N_rec_filtered
+        num_c_boxes += 2.0
+
+    # There are a bunch of 1 and 2D interpolation tables, but ignore those as they are small relative to 3D grids
+
+    if flag_options.USE_MASS_DEPENDENT_ZETA and flag_options.USE_MINI_HALOS:
+        # log10_Mturnover_unfiltered, log10_Mturnover_filtered, log10_Mturnover_MINI_unfiltered, log10_Mturnover_MINI_filtered
+        num_c_boxes += 4.0
+
+    if flag_options.USE_HALO_FIELD:
+        # M_coll_unfiltered, M_coll_filtered
+        num_c_boxes += 2.0
+
+    # These are all fftwf complex arrays (thus 2 * size)
+    size_c = (2.0 * (np.float32(1.0).nbytes)) * num_c_boxes * hii_kspace_num_pixels
+
+    return {"python": size_py, "c": size_c}
+
+
+def mem_spin_temperature(
+    *,
+    user_params=None,
+    astro_params=None,
+    flag_options=None,
+):
+    """A function to estimate total memory usage of a spin_temperature call."""
+    """Memory usage of Python IonizedBox class."""
+
+    """All declared HII_DIM boxes"""
+    # Ts_bx, x_e_box, Tk_box, J_21_LW_box
+    num_py_boxes = 4.0
+
+    size_py = num_py_boxes * (user_params.HII_DIM) ** 3
+
+    # These are all float arrays
+    size_py = (np.float32(1.0).nbytes) * size_py
+
+    """Memory usage within SpinTemperatureBox.c"""
+    hii_kspace_num_pixels = (float(user_params.HII_DIM) / 2.0 + 1.0) * (
+        user_params.HII_DIM
+    ) ** 2
+
+    # box, unfiltered_box
+    num_c_boxes = 2.0
+    num_c_boxes_alt = 0.0
+    num_c_boxes_initialised = 0.0
+
+    mem_c_interp = 0.0
+
+    if flag_options.USE_MINI_HALOS:
+        # log10_Mcrit_LW_unfiltered, log10_Mcrit_LW_filtered
+        num_c_boxes += 2.0
+
+        # log10_Mcrit_LW
+        num_c_boxes_alt += global_params.NUM_FILTER_STEPS_FOR_Ts
+
+    # There are a bunch of 1 and 2D interpolation tables, but ignore those as they are small relative to 3D grids
+    # I will add any that could end up considerably large.
+
+    if flag_options.USE_MASS_DEPENDENT_ZETA:
+        # delNL0
+        num_c_boxes_initialised += global_params.NUM_FILTER_STEPS_FOR_Ts
+
+        # del_fcoll_Rct, m_xHII_low_box, inverse_val_box
+        num_c_boxes_initialised += (
+            3.0  # m_xHII_low_box is an int, but I'll count it as 4 bytes just in case
+        )
+
+        # dxheat_dt_box, dxion_source_dt_box, dxlya_dt_box, dstarlya_dt_box
+        num_c_boxes_initialised += 2.0 * 4.0  # factor of 2. as these are doubles
```

Comment:
shall we reduce it to float32 then?

Reply:
No, these should stay as double for numerical accuracy